File: 20-actors/kabuto/methods/test_social.py
#!/usr/bin/env python3
"""Tests for the kabuto 兜 social composer + Charter-Rider gate (methods/social.py).

    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest test_social.py
    python3 test_social.py

Covers the Charter-Rider §2(a)-(h) deny-scan (the content gate every post must pass) and
the G2 framing invariant: every composed post is aggregate-first, public-facts-only, and
framed as resilience/accountability — never a target-list.
"""
from __future__ import annotations

import pathlib
import sys

try:
    from kabuto_edn import classify, load_edn
    from social import charter_rider_clean, compose, post_record
except ImportError:
    from kabuto.methods.kabuto_edn import classify, load_edn  # type: ignore
    from kabuto.methods.social import charter_rider_clean, compose, post_record  # type: ignore

_SEED = pathlib.Path(__file__).resolve().parent.parent / "data" / "seed-public-companies.kotoba.edn"


def test_charter_rider_rejects_prohibited_content():
    # §2(a) force / §2 weaponisation-adjacent phrasing must be refused
    assert charter_rider_clean("where to cut the supply chain") is False
    assert charter_rider_clean("weapon design for the foundry") is False
    assert charter_rider_clean("buy adsense placement") is False


def test_charter_rider_accepts_clean_resilience_text():
    assert charter_rider_clean(
        "Disclosed supply dependency — diversify to build resilience.") is True


def test_compose_posts_pass_the_rider_gate():
    rows = load_edn(_SEED)
    companies, _a, _c, edges, _p = classify(rows)
    posts = list(compose(companies, edges, "top jurisdictions US/JP/CN"))
    assert posts, "expected composed posts"
    for _subject, _kind, text in posts:
        assert charter_rider_clean(text), f"composed post violates the Rider: {text!r}"


def test_compose_headline_is_not_a_target_list():
    rows = load_edn(_SEED)
    companies, _a, _c, edges, _p = classify(rows)
    posts = list(compose(companies, edges, "summary"))
    headline = [t for _s, kind, t in posts if kind == "intel-report"]
    assert headline and "not a target-list" in headline[0]


def test_compose_supply_edges_frame_diversification():
    rows = load_edn(_SEED)
    companies, _a, _c, edges, _p = classify(rows)
    edge_posts = [t for _s, kind, t in compose(companies, edges, "") if kind == "supply-edge"]
    assert edge_posts
    # the actionable framing is diversification/resilience, never interdiction
    for t in edge_posts:
        assert "resilience" in t.lower() or "diversify" in t.lower()


def test_post_record_is_well_formed_atproto_post():
    rec = post_record("hello", ["en"])
    assert rec["$type"] == "app.bsky.feed.post"
    assert len(rec["text"]) <= 300 and rec["createdAt"].endswith("Z")


def _run():
    fns = [v for k, v in sorted(globals().items())
           if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"kabuto social.py: {len(fns)}/{len(fns)} tests passed")
    return True


if __name__ == "__main__":
    sys.exit(0 if _run() else 1)

File: 20-actors/kabuto/methods/test_social.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_social.py.
(ns kabuto.methods.test-social
  "Tests for the kabuto 兜 social composer + Charter-Rider gate (methods/social.clj).

  Covers the Charter-Rider §2(a)-(h) deny-scan + the G2 framing invariant: every composed post
  is aggregate-first, public-facts-only, framed as resilience/accountability — never a target-list.

  Run:  bb --classpath 20-actors 20-actors/kabuto/methods/test_social.clj"
  (:require [kabuto.methods.kabuto-edn :as e]
            [kabuto.methods.social :as s]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [clojure.test :refer [deftest is run-tests]]))

(def ^:private this-file *file*)
(defn- g []
  (e/classify (e/load-edn (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
                              (io/file "data" "seed-public-companies.kotoba.edn")))))

(deftest charter-rider-rejects-prohibited-content
  (is (false? (s/charter-rider-clean "where to cut the supply chain")))
  (is (false? (s/charter-rider-clean "weapon design for the foundry")))
  (is (false? (s/charter-rider-clean "buy adsense placement"))))

(deftest charter-rider-accepts-clean-resilience-text
  (is (true? (s/charter-rider-clean "Disclosed supply dependency — diversify to build resilience."))))

(deftest compose-posts-pass-the-rider-gate
  (let [gr (g)
        posts (s/compose (:companies gr) (:edges gr) "top jurisdictions US/JP/CN")]
    (is (seq posts) "expected composed posts")
    (doseq [[_ _ text] posts]
      (is (s/charter-rider-clean text) (str "composed post violates the Rider: " text)))))

(deftest compose-headline-is-not-a-target-list
  (let [gr (g)
        posts (s/compose (:companies gr) (:edges gr) "summary")
        headline (first (keep (fn [[_ kind t]] (when (= kind "intel-report") t)) posts))]
    (is (and headline (str/includes? headline "not a target-list")))))

(deftest compose-supply-edges-frame-diversification
  (let [gr (g)
        edge-posts (keep (fn [[_ kind t]] (when (= kind "supply-edge") t))
                         (s/compose (:companies gr) (:edges gr) ""))]
    (is (seq edge-posts))
    (doseq [t edge-posts]
      (is (or (str/includes? (str/lower-case t) "resilience")
              (str/includes? (str/lower-case t) "diversify"))))))

(deftest post-record-is-well-formed-atproto-post
  (let [rec (s/post-record "hello" ["en"])]
    (is (= (get rec "$type") "app.bsky.feed.post"))
    (is (<= (count (get rec "text")) 300))
    (is (str/ends-with? (get rec "createdAt") "Z"))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'kabuto.methods.test-social)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
